lib/active_merchant/billing/gateways/payeezy.rb

```diff
@@ -140,6 +142,25 @@ def scrub(transcript)
 
       private
 
+      def add_external_three_ds(params, payment_method, options)
+        return if options[:three_d_secure].blank?
+
+        three_d_secure_data = {}
+        three_d_secure_data[:program_protocol] = options[:three_d_secure][:version] if options[:three_d_secure][:version].present?
+        three_d_secure_data[:directory_server_transaction_id] = options[:three_d_secure][:ds_transaction_id] if options[:three_d_secure][:ds_transaction_id].present?
+        three_d_secure_data[:cardholder_name] = payment_method.name
+        three_d_secure_data[:card_number] = payment_method.number
+        three_d_secure_data[:exp_date] = format_exp_date(payment_method.month, payment_method.year)
+        three_d_secure_data[:cvv] = payment_method.verification_value if payment_method.verification_value?
+        three_d_secure_data[:xid] = options[:three_d_secure][:acs_transaction_id]
+        three_d_secure_data[:cavv] = options[:three_d_secure][:cavv]
+        three_d_secure_data[:wallet_provider_id] = 'NO_WALLET'
+        three_d_secure_data[:type] = 'D'
+        params[:eci_indicator] = options[:three_d_secure][:eci]
+        params[:method] = '3DS'
+        params['3DS'.to_sym] = three_d_secure_data
+      end
+
```

Comment:
💭 Suggestion.

Maybe a variable thre_d_secure = options[:three_d_secure], avoid using options[:three_d_secure].

Reply:
with the Cristian refactor I think this is solved.